### PR TITLE
[USER32] Hackfix for CORE-17902 - The cursoricon (copyimage) that returns NULL to render - LR_COPYFROMRESOURCE

### PR DESCRIPTION
**Pull request by julenuri**
Hackfix for CORE-17902 - The cursoricon (copyimage) that returns NULL to render - LR_COPYFROMRESOURCE

## Purpose

In this case, the aim is to render some VB5/6 software icons to show up. Those elements are icons rendered in a way that is common under Visual Basic icons loaded from a .ocx .

ThunderRT6FormDC->ThunderRT6UserControlDC->RebarWindows32->Toolbar20WndClass->msvb_lib_toolbar

Quoting Simone:

> Unable to render embedded object: File (ReactOS_workaround) not found.
> 
> it is making a CopyImage call with the LR_COPYFROMRESOURCE|LB_COPYRETURNORG flags. All of these are returning NULL.
> LB_COPYRETURNORG flag is only managed by bitmap copy function (BITMAP_CopyImage), not by the cursoricon copy function (CURSORICON_CopyImage) but it's not the critical one.
> The critical flag is the LR_COPYFROMRESOURCE. 

More info here: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-copyimage

>By making it to return the original bitmap in the case of failure, makes the icon shown as expected.
>The issue so resides on application of api ("tries to" means "tries to" so the fallback to return the original image isneeded), or there is a failure on child NtUserGetIconInfo.

- Fix JIRA issue: [CORE-17902](https://jira.reactos.org/browse/CORE-17902)

## Proposed changes

Minor (fix? or hackfix?) that should make it to work. 
 
BEFORE:
Before

AFTER: 
After

## TODO

- [ ] Investigate if the behavior is correct comparing to Windows.
- [ ] Investigate if there is any better way to implement it/or if there are similar failures under a more simple software to test.

### CODE REVIEW

File: win32ss/user/user32/windows/cursoricon.c [hunk 1]
```diff
@@ -2025,13 +2025,17 @@ HANDLE WINAPI CopyImage(
...
     {
         case IMAGE_BITMAP:
             return BITMAP_CopyImage(hImage, cxDesired, cyDesired, fuFlags);
         case IMAGE_CURSOR:
         case IMAGE_ICON:
-            return CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
+            handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
+            if(handle == NULL && (fuFlags & (LR_COPYFROMRESOURCE|LR_COPYRETURNORG)))
+                handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, (fuFlags & ~LR_COPYFROMRESOURCE));
+            return handle;
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2038]:
```suggestion
        {
            HANDLE handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
            if (!handle && (fuFlags & (LR_COPYFROMRESOURCE|LR_COPYRETURNORG)))
                handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, (fuFlags & ~LR_COPYFROMRESOURCE));
            return handle;
        }
```

File: win32ss/user/user32/windows/cursoricon.c [hunk 2]
```diff
@@ -2025,13 +2025,17 @@ HANDLE WINAPI CopyImage(
 {
     TRACE("hImage=%p, uType=%u, cxDesired=%d, cyDesired=%d, fuFlags=%x\n",
         hImage, uType, cxDesired, cyDesired, fuFlags);
+    HANDLE handle;
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2028]:
```suggestion

```

File: win32ss/user/user32/windows/cursoricon.c [hunk 3]
```diff
@@ -2025,13 +2025,17 @@ HANDLE WINAPI CopyImage(
...
     switch(uType)
     {
         case IMAGE_BITMAP:
             return BITMAP_CopyImage(hImage, cxDesired, cyDesired, fuFlags);
         case IMAGE_CURSOR:
         case IMAGE_ICON:
-            return CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
+            handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
+            if(handle == NULL && (fuFlags & (LR_COPYFROMRESOURCE|LR_COPYRETURNORG)))
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2036]: When you run the VB-based app in ReactOS, which flag(s) amongst these two are being specified? Both? or only one of these?

Comment (HBelusca) [reply win32ss/user/user32/windows/cursoricon.c:2036]: In addition I would suggest you add here in the code a `// HACK` or `// HACKFIX` comment explaining what's happening, with the CORE-xxxx number mentioned too.

I also wonder whether this hack could be added instead into the `CURSORICON_CopyImage` function instead? In any case I guess running the test bots on that would be necessary.

File: win32ss/user/user32/windows/cursoricon.c (hunk 1 above)
Comment (julenuri) [reply win32ss/user/user32/windows/cursoricon.c:2038]: IIRC: That first patch failed for me under GCC (is it before C99?):
```
../win32ss/user/user32/windows/cursoricon.c: In function 'CopyImage':
../win32ss/user/user32/windows/cursoricon.c:2034:13: error: a label can only be part of a statement and a declaration is not a statement
             HANDLE handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
             ^~
```
But I have to check.

Comment (HBelusca) [reply win32ss/user/user32/windows/cursoricon.c:2038]: yes that's why you need the curly brackets for the case XXX: , as in my suggestion.

File: win32ss/user/user32/windows/cursoricon.c (hunk 3 above)
Comment (julenuri) [reply win32ss/user/user32/windows/cursoricon.c]: Both are being specified according to the ticket info. LR_COPYFROMRESOURCE and LB_COPYRETURNORG.

Not sure about the second comment. 
I do know that under Wine, it always show up the icons, but not sure how they do. And not 100% sure how Windows does at all. So if anyone can bring light, better than better.
I think I can't execute the test bot by myself, I have no rights for that I guess.

**Comment by Doug-Lyons** [comment]
Testbot Results:
refs/pull/6886/merge on top of 0.4.15-dev-8171-gc97c1ad

KVM:  https://reactos.org/testman/compare.php?ids=95854,95861 LGTM
VBox: https://reactos.org/testman/compare.php?ids=95853,95862 LGTM

File: win32ss/user/user32/windows/cursoricon.c [hunk 4]
```diff
@@ -2031,7 +2031,15 @@ HANDLE WINAPI CopyImage(
             return BITMAP_CopyImage(hImage, cxDesired, cyDesired, fuFlags);
         case IMAGE_CURSOR:
         case IMAGE_ICON:
-            return CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
+     /* HACK: Copying bitmaps with LR_COPYFROMRESOURCE flag fails. CORE-17902.
+     * This is a way to return original bit map it if we want
```
Comment (Doug-Lyons) [inline win32ss/user/user32/windows/cursoricon.c:2035]: Suggestion:
```
 * This is a way to return the original bit map if we want
```

**Review by Doug-Lyons — APPROVED** [review]

**Comment by Doug-Lyons** [comment]
I was able to test this and the icons show up after making these changes.

File: win32ss/user/user32/windows/cursoricon.c [hunk 5]
```diff
@@ -2031,7 +2031,15 @@ HANDLE WINAPI CopyImage(
...
+     * the icons to show up. We need a simpler test. */
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2036]:
```suggestion
        /* HACK: Copying bitmaps with LR_COPYFROMRESOURCE flag fails. CORE-17902.
         * This is a way to return the original bit map if we need
         * the icons to show up. We need a simpler test. */
```

**Review by oleg-dubinskiy — APPROVED** [review]
It seems to be not a proper solution, but it's OK as a temporary workaround, until the correct fix is made.

**Review by JoachimHenze — APPROVED** [review]
why not... `Fake it till you make it(TM)!` Approved.

**Pull request merged**